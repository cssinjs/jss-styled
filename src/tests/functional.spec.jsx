import 'react-dom'
import React from 'react'
import Observable from 'zen-observable'
import Enzyme, {mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import {
  getCss,
  removeWhitespace
} from './utils'

import CreateApp from './App'

Enzyme.configure({adapter: new Adapter()})

let Styled
let styled
let ThemeProvider

const mockNameGenerators = () => {
  let styledCounter = 0

  jest.mock('../utils/generateTagName')
  jest.mock('jss/lib/utils/createGenerateClassName')

  const generateTagName = require('../utils/generateTagName').default
  const createGenerateClassName = require('jss/lib/utils/createGenerateClassName').default

  // $FlowIgnore there is now mockImplementation in declaration
  generateTagName.mockImplementation((tagName: string) => `${tagName}-${++styledCounter}`)
  createGenerateClassName.mockImplementation(() => rule => `${rule.key}-id`)
}

const assertSheet = (sheet) => {
  expect(sheet.toString()).toMatchSnapshot()
  expect(getCss(sheet)).toBe(removeWhitespace(sheet.toString()))
}

const assertComponent = (wrapper) => {
  expect(wrapper).toMatchSnapshot()
}

describe('functional tests', () => {
  beforeEach(() => {
    mockNameGenerators()

    ;({ThemeProvider, Styled} = require('../'))
    styled = Styled()
  })

  afterEach(() => {
    styled.sheet.detach()
  })

  it('should update props and unmount', () => {
    const App = CreateApp(styled)
    const wrapper = mount(<App />)
    const {sheet} = styled

    assertSheet(sheet)
    wrapper.setProps({margin: 20})
    assertSheet(sheet)
    wrapper.unmount()
  })

  /**
   * TODO: we should return this test when an issue with nesting order will be resolved
   * @see https://github.com/cssinjs/jss/pull/655
   */

  it.skip('should update nested props', () => {
    styled = Styled({
      button: {
        fontSize: 12,
      },
    })

    const Div = styled('div')({
      padding: 15,
      '&:hover': {
        '& $button': {
          color: ({primary}) => (primary ? 'red' : 'green'),
        },
      },
    })

    const Button = styled('button')({
      composes: '$button',
    })

    const App = (props: {primary?: boolean}) => (
      <div>
        <Div primary={props.primary}>
          <Button>Button</Button>
        </Div>
        <Div primary={!props.primary}>
          <Button>Button</Button>
        </Div>
      </div>
    )

    const wrapper = mount(<App />)
    const {sheet} = styled

    assertSheet(sheet)
    wrapper.setProps({primary: true})
    assertSheet(sheet)
    wrapper.unmount()
  })

  it('should use props on remount', () => {
    const Button = styled('button')({
      color: props => (props.primary ? 'red' : 'black')
    })

    const wrapper = mount(<Button />)
    const {sheet} = styled

    assertSheet(sheet)
    wrapper
      .unmount()
      .mount()
      .setProps({primary: true})
    assertSheet(sheet)
    wrapper.unmount()
  })

  it('should update dynamic props for conditional rules', () => {
    const Button = styled('button')({
      padding: props => (props.spaced ? 10 : 0),

      '@media screen': {
        '& .button': {
          margin: props => (props.spaced ? 10 : 0)
        }
      }
    })

    const wrapper = mount(<Button spaced />)
    const {sheet} = styled

    assertSheet(sheet)
    wrapper.setProps({spaced: false})
    assertSheet(sheet)

    wrapper.unmount()
  })

  it('should use Styled Component classname in string', () => {
    const AuthorName = styled('div')({color: 'red'})
    const Avatar = styled('img')({width: props => props.width})

    const Message = styled('div')({
      [`&:not(:first-child) ${AuthorName}`]: {
        display: 'none'
      },
      [`&:not(:last-child) ${Avatar}`]: {
        visibility: 'hidden'
      },
      [`& ${AuthorName}`]: {
        color: 'green'
      }
    })

    assertComponent(mount(
      <Message>
        <AuthorName>name</AuthorName>
        <Avatar width={30} />
      </Message>
    ))

    assertSheet(styled.sheet)
  })

  describe('Compose React Components', () => {
    it('should use .name', () => {
      const Test = props => <h1 {...props}>test</h1>
      const StyledTest = styled(Test)({
        padding: 10,
      })
      assertComponent(mount(<StyledTest />))
      assertSheet(styled.sheet)
    })

    it('should use .displayName', () => {
      const Test = props => <h1 {...props}>test</h1>
      Test.displayName = 'TestDisplayName'
      const StyledTest = styled(Test)({
        padding: 10,
      })
      assertComponent(mount(<StyledTest />))
      assertSheet(styled.sheet)
    })

    it('should escape name in dev mode', () => {
      const Comp = ({className}: {className: string}) => (
        <div className={className}>Container</div>
      )

      Comp.displayName = '(Comp.name)'

      const Container = styled(Comp)({
        color: 'red',
      })

      const wrapper = mount(<Container />)
      const {sheet} = styled

      assertSheet(sheet)

      wrapper.unmount()
    })

    it('should use .name', () => {
      const StyledTest = styled(props => <h1 {...props}>test</h1>)({
        padding: 10,
      })
      assertComponent(mount(<StyledTest />))
      assertSheet(styled.sheet)
    })

    it('should pass props', () => {
      const StyledTest = styled(props => JSON.stringify(props))({
        padding: 10,
      })
      assertComponent(mount(
        <StyledTest testProp={1} testProp2="2" className="testClassName" />
      ))
      assertSheet(styled.sheet)
    })
  })

  describe('Observables', () => {
    it('should use observable value', () => {
      let observer

      const Container = styled('div')({
        padding: 40,
        height: new Observable((o) => {
          observer = o
        }),
        textAlign: 'center'
      })

      const wrapper = mount(<Container />)
      const {sheet} = styled

      if (observer) {
        observer.next('10px')
        assertSheet(sheet)
      }
      else {
        throw new Error('there is no observable value')
      }


      wrapper.unmount()
    })
  })

  describe('composable styles', () => {
    it('should merge all object styles', () => {
      const Button = styled('button')(
        {color: props => props.color},
        {margin: 10},
        {padding: 20},
        {backgroundColor: props => props.backgroundColor}
      )

      const wrapper = mount(
        <Button color="red" backgroundColor="green" />
      )

      const {sheet} = styled

      assertSheet(sheet)
      wrapper.unmount()
    })

    it('should compose all function styles', () => {
      const Button = styled('button')(
        (props => props.theme === 'action' && ({
          color: 'red',
          backgroundColor: 'green'
        })),
        (props => props.theme === 'normal' && ({
          color: 'white',
          backgroundColor: 'black'
        })),
        (() => ({
          margin: 20,
          padding: 20
        }))
      )

      const wrapper = mount(
        <Button theme="action" />
      )
      const {sheet} = styled

      assertSheet(sheet)
      wrapper.setProps({theme: 'normal'})
      assertSheet(sheet)
      wrapper.unmount()
    })

    it('should merge and compose all styles', () => {
      const round = props => props.round && ({
        'border-radius': '100%'
      })

      const theme = props => ({
        action: {
          color: 'red',
          backgroundColor: 'green',
        },
        normal: {
          color: 'black',
          backgroundColor: 'white',
        },
      })[props.theme]

      const Button = styled('button')({
        margin: 20,
        padding: 20,
        fontWeight: () => 400,
      }, round, theme, {fontSize: ({fontSize}) => fontSize})

      const wrapper = mount(
        <Button theme="action" round />
      )
      const {sheet} = styled

      assertComponent(wrapper)

      assertSheet(sheet)
      wrapper.setProps({theme: 'normal', fontSize: 15, round: false})
      assertSheet(sheet)
      wrapper.unmount()
    })
  })

  describe('theming', () => {
    it('should work with ThemeProvider', () => {
      const initialTheme = {
        color: {
          primary: 'green',
          secondary: 'white'
        }
      }

      const Button = styled('button')(({theme}) => ({
        color: theme.color.primary,
        'background-color': theme.color.secondary,
      }))

      const wrapper = mount(
        <ThemeProvider theme={initialTheme}>
          <Button />
        </ThemeProvider>
      )
      const {sheet} = styled

      assertSheet(sheet)

      wrapper.unmount()
    })

    it('should update theme', () => {
      const initialTheme = {
        color: {
          primary: 'green',
          secondary: 'white'
        }
      }

      const Button = styled('button')(({theme}) => ({
        color: theme.color.primary,
        'background-color': theme.color.secondary,
      }))

      const App = (props: {theme: Object}) => (
        <ThemeProvider theme={props.theme}>
          <Button />
        </ThemeProvider>
      )

      const wrapper = mount(<App theme={initialTheme} />)
      const {sheet} = styled

      assertSheet(sheet)

      const nextTheme = {
        color: {
          primary: 'yellow',
          secondary: 'blue'
        }
      }
      wrapper.setProps({theme: nextTheme})

      assertSheet(sheet)

      wrapper.unmount()
    })

    it('should work with nested ThemeProvider', () => {
      const themes = [{
        color: {
          primary: 'green',
          secondary: 'white'
        }
      }, {
        color: {
          primary: 'blue',
          secondary: 'yellow'
        }
      }]

      const Button = styled('button')(({theme}) => ({
        color: theme.color.primary,
        'background-color': theme.color.secondary,
      }))

      const App = () => (
        <ThemeProvider theme={themes[0]}>
          <div>
            <Button />
            <ThemeProvider theme={themes[1]}>
              <Button />
            </ThemeProvider>
          </div>
        </ThemeProvider>
      )

      const wrapper = mount(<App />)
      const {sheet} = styled

      assertSheet(sheet)

      wrapper.unmount()
    })
  })
})
