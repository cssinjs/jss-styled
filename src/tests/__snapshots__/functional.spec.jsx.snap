// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functional tests Compose React Components should escape name in dev mode 1`] = `
".\\\\(Comp\\\\.name\\\\)-1-id {
  color: red;
}"
`;

exports[`functional tests Compose React Components should pass props 1`] = `
<StyledElement
  className="testClassName"
  testProp={1}
  testProp2="2"
>
  <Component
    className="StyledElement-1-id testClassName"
    testProp={1}
    testProp2="2"
  >
    {"testProp":1,"testProp2":"2","className":"StyledElement-1-id testClassName"}
  </Component>
</StyledElement>
`;

exports[`functional tests Compose React Components should pass props 2`] = `
".StyledElement-1-id {
  padding: 10px;
}"
`;

exports[`functional tests Compose React Components should use .displayName 1`] = `
<StyledElement>
  <TestDisplayName
    className="TestDisplayName-1-id"
  >
    <h1
      className="TestDisplayName-1-id"
    >
      test
    </h1>
  </TestDisplayName>
</StyledElement>
`;

exports[`functional tests Compose React Components should use .displayName 2`] = `
".TestDisplayName-1-id {
  padding: 10px;
}"
`;

exports[`functional tests Compose React Components should use .name 1`] = `
<StyledElement>
  <Test
    className="Test-1-id"
  >
    <h1
      className="Test-1-id"
    >
      test
    </h1>
  </Test>
</StyledElement>
`;

exports[`functional tests Compose React Components should use .name 2`] = `
".Test-1-id {
  padding: 10px;
}"
`;

exports[`functional tests Compose React Components should use .name 3`] = `
<StyledElement>
  <Component
    className="StyledElement-1-id"
  >
    <h1
      className="StyledElement-1-id"
    >
      test
    </h1>
  </Component>
</StyledElement>
`;

exports[`functional tests Compose React Components should use .name 4`] = `
".StyledElement-1-id {
  padding: 10px;
}"
`;

exports[`functional tests Observables should use observable value 1`] = `
".div-1-id {
  padding: 40px;
  text-align: center;
}
.div-2-id {
  height: 10px;
}"
`;

exports[`functional tests composable styles should compose all function styles 1`] = `
".button-1-id {
  color: red;
  background-color: green;
  margin: 20px;
  padding: 20px;
}"
`;

exports[`functional tests composable styles should compose all function styles 2`] = `
".button-1-id {
  color: white;
  background-color: black;
  margin: 20px;
  padding: 20px;
}"
`;

exports[`functional tests composable styles should merge all object styles 1`] = `
".button-1-id {
  margin: 10px;
  padding: 20px;
}
.button-2-id {
  color: red;
  background-color: green;
}"
`;

exports[`functional tests composable styles should merge and compose all styles 1`] = `
<StyledElement
  round={true}
  theme="action"
>
  <button
    className="button-1-id button-2-id button-3-id"
  />
</StyledElement>
`;

exports[`functional tests composable styles should merge and compose all styles 2`] = `
".button-1-id {
  margin: 20px;
  padding: 20px;
}
.button-2-id {
  font-weight: 400;
}
.button-3-id {
  border-radius: 100%;
  color: red;
  background-color: green;
}"
`;

exports[`functional tests composable styles should merge and compose all styles 3`] = `
".button-1-id {
  margin: 20px;
  padding: 20px;
}
.button-2-id {
  font-weight: 400;
  font-size: 15px;
}
.button-3-id {
  color: black;
  background-color: white;
}"
`;

exports[`functional tests should update dynamic props for conditional rules 1`] = `
".button-1-id {
  padding: 10px;
}
@media screen {
  .button-1-id {
  }
  .button-1-id .button {
    margin: 10px;
  }
}"
`;

exports[`functional tests should update dynamic props for conditional rules 2`] = `
".button-1-id {
  padding: 0;
}
@media screen {
  .button-1-id {
  }
  .button-1-id .button {
    margin: 0;
  }
}"
`;

exports[`functional tests should update nested props 1`] = `
".button-id {
  font-size: 12px;
}
.div-1-id {
  padding: 15px;
}
.div-3-id {
}
.div-3-id:hover {
}
.div-3-id:hover .button-id {
  color: green;
}
.button-2-id {
}
.div-4-id {
}
.div-4-id:hover {
}
.div-4-id:hover .button-id {
  color: red;
}"
`;

exports[`functional tests should update nested props 2`] = `
".button-id {
  font-size: 12px;
}
.div-1-id {
  padding: 15px;
}
.div-3-id:hover .button-id {
  color: red;
}
.div-4-id:hover .button-id {
  color: green;
}"
`;

exports[`functional tests should update props and unmount 1`] = `
".div-1-id {
  margin: 50px;
}
.header-2-id {
  padding: 10px;
}
.h1-5-id {
  color: red;
}
.section-3-id {
  color: red;
}
.button-6-id {
  margin: 0;
}
.button-7-id {
  margin: 10px;
}
.section-4-id {
  color: yellow;
}"
`;

exports[`functional tests should update props and unmount 2`] = `
".div-1-id {
  margin: 50px;
}
.header-2-id {
  padding: 10px;
}
.h1-5-id {
  color: red;
}
.section-3-id {
  color: red;
}
.button-6-id {
  margin: 0;
}
.button-7-id {
  margin: 20px;
}
.section-4-id {
  color: yellow;
}"
`;

exports[`functional tests should use Styled Component classname in string 1`] = `
<StyledElement>
  <div
    className="div-4-id"
  >
    <StyledElement>
      <div
        className="static-2-id div-1-id"
      >
        name
      </div>
    </StyledElement>
    <StyledElement
      width={30}
    >
      <img
        className="static-3-id img-5-id"
        width={30}
      />
    </StyledElement>
  </div>
</StyledElement>
`;

exports[`functional tests should use Styled Component classname in string 2`] = `
".div-4-id {
}
.div-4-id:not(:first-child) .static-2-id {
  display: none;
}
.div-4-id:not(:last-child) .static-3-id {
  visibility: hidden;
}
.div-4-id .static-2-id {
  color: green;
}
.div-1-id {
  color: red;
}
.img-5-id {
  width: 30px;
}"
`;

exports[`functional tests should use props on remount 1`] = `
".button-1-id {
  color: black;
}"
`;

exports[`functional tests should use props on remount 2`] = `
".button-2-id {
  color: red;
}"
`;

exports[`functional tests theming should update theme 1`] = `
".button-1-id {
  color: green;
  background-color: white;
}"
`;

exports[`functional tests theming should update theme 2`] = `
".button-1-id {
  color: yellow;
  background-color: blue;
}"
`;

exports[`functional tests theming should work with ThemeProvider 1`] = `
".button-1-id {
  color: green;
  background-color: white;
}"
`;

exports[`functional tests theming should work with nested ThemeProvider 1`] = `
".button-1-id {
  color: green;
  background-color: white;
}
.button-2-id {
  color: blue;
  background-color: yellow;
}"
`;
